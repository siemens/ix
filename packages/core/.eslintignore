playwright-ct.config.ts
playwright.config.ts
