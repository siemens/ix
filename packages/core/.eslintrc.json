{
  "root": true,
  "parserOptions": {
    "project": "./tsconfig.json"
  },
  "extends": ["plugin:@stencil-community/recommended", "ix"],
  "rules": {
    "@stencil-community/async-methods": 0,
    "@stencil-community/own-props-must-be-private": 0,
    "@stencil-community/own-methods-must-be-private": 0,
    "@stencil-community/strict-boolean-conditions": 0,
    "@stencil-community/ban-default-true": ["warn"],
    "react/jsx-no-bind": 0,
    "react/jsx-uses-react": 0,
    "react/react-in-jsx-scope": 0
  }
}
