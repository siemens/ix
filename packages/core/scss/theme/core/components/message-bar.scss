@mixin setVars {
  --theme-message-bar--border-thickness: var(--theme-border-width-thick);
  --theme-message-bar--border-radius: var(--theme-default-border-radius);
}