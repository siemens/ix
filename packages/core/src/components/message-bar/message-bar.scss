/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@import 'common-variables';
@import 'mixins/text-truncation';
@import 'mixins/shadow-dom/component';

:host {
  @include ix-component;

  .message-container {
    display: flex;
    flex-direction: row;
    align-items: flex-start;
    flex-wrap: nowrap;
    justify-content: space-between;
    min-height: 3.375rem;
    margin: $small-space $small-space 0rem $small-space;
    padding: calc(#{$medium-space} - var(--theme-message-bar--border-thickness))
      $medium-space
      calc(#{$medium-space} - var(--theme-message-bar--border-thickness))
      $default-space;
    border-radius: var(--theme-message-bar--border-radius);
    background-color: var(--theme-messagebar--background);
  }

  .danger {
    border: solid var(--theme-message-bar--border-thickness)
      var(--theme-color-alarm);
  }

  .warning {
    border: solid var(--theme-message-bar--border-thickness)
      var(--theme-color-warning);
  }

  .info {
    border: solid var(--theme-message-bar--border-thickness)
      var(--theme-color-info);
  }

  .message-content {
    @include ellipsis;
    flex-grow: 1;
    align-self: center;
    min-height: 1.25rem;
    padding: 0 $default-space;
    font-weight: bold;
    white-space: normal;
  }

  ix-icon {
    margin-top: $tiny-space;
  }
}
