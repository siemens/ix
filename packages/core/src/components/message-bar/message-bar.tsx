/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  Component,
  Event,
  EventEmitter,
  h,
  Host,
  Prop,
  State,
} from '@stencil/core';
import anime from 'animejs';
import { NotificationColor } from '../utils/notification-color';

@Component({
  tag: 'ix-message-bar',
  styleUrl: 'message-bar.scss',
  shadow: true,
})
export class MessageBar {
  /**
   * Specifies the type of the alert.
   */
  @Prop() type: 'danger' | 'warning' | 'info' = 'info';

  /**
   * If true, close button is enabled and alert can be dismissed by the user
   */
  @Prop() dismissible = true;

  /**
   * An event emitted when the close button is clicked
   */
  @Event() closedChange: EventEmitter;

  @State() icon: 'error' | 'warning' | 'info';

  @State() color: NotificationColor;

  private static readonly duration = 300;

  private divElement?: HTMLElement;

  componentWillRender() {
    if (this.type === 'danger') {
      this.icon = 'error';
      this.color = 'color-alarm';
    }

    if (this.type === 'info') {
      this.icon = 'info';
      this.color = 'color-info';
    }

    if (this.type === 'warning') {
      this.icon = 'warning';
      this.color = 'color-warning';
    }
  }

  private closeAlert(el: HTMLElement) {
    anime({
      targets: el,
      duration: MessageBar.duration,
      opacity: [1, 0],
      easing: 'easeOutSine',
      complete: () => {
        el.classList.add('d-none');
      },
    });
    this.closedChange.emit();
  }

  render() {
    return (
      <Host>
        <div
          class={{ 'message-container': true, [this.type]: true }}
          role="alert"
          ref={(el) => (this.divElement = el as HTMLElement)}
        >
          <ix-icon color={this.color} name={this.icon}></ix-icon>
          <div class="message-content">
            <slot></slot>
          </div>
          {this.dismissible ? (
            <ix-icon-button
              icon="close"
              size="24"
              ghost={true}
              onClick={() => {
                this.closeAlert(this.divElement);
              }}
              data-testid="close-btn"
            ></ix-icon-button>
          ) : (
            ''
          )}
        </div>
      </Host>
    );
  }
}
