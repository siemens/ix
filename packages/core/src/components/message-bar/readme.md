<!--
SPDX-FileCopyrightText: 2022 Siemens AG

SPDX-License-Identifier: MIT
-->

# message-bar

<!-- Auto Generated Below -->

## Properties

| Property      | Attribute     | Description                                                             | Type                              | Default  |
| ------------- | ------------- | ----------------------------------------------------------------------- | --------------------------------- | -------- |
| `dismissible` | `dismissible` | If true, close button is enabled and alert can be dismissed by the user | `boolean`                         | `true`   |
| `type`        | `type`        | Specifies the type of the alert.                                        | `"danger" \| "info" \| "warning"` | `'info'` |

## Events

| Event          | Description                                       | Type               |
| -------------- | ------------------------------------------------- | ------------------ |
| `closedChange` | An event emitted when the close button is clicked | `CustomEvent<any>` |

---
