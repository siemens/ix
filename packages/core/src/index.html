<!--
SPDX-FileCopyrightText: 2023 Siemens AG

SPDX-License-Identifier: MIT
-->

<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Stencil Component Starter</title>

  <script type="module" src="/build/siemens-ix.esm.js"></script>
  <script nomodule src="/build/siemens-ix.js"></script>

  <link rel="stylesheet" href="/build/siemens-ix.css" />
  <link rel="stylesheet" href="/build/ix-brand-theme/dist/ix-brand-theme/ix-brand-theme.css" />
  <link rel="stylesheet" href="/build/ix-icons/css/ix-icons.css" />
  <script type="module">
    import { defineCustomElements } from '/build/ix-brand-theme/loader/index.es2017.js';
    defineCustomElements();
  </script>
</head>

<body style="margin: 0px; height: 100vh">
  <my-component />
</body>

</html>