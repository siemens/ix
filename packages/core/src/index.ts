/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import '@siemens/ix-icons';
import type {
  Components as IxIconsComponents,
  JSX as IxIconsJSX,
} from '@siemens/ix-icons';
export * from './components';
export * from './components/category-filter/filter-state';
export * from './components/category-filter/input-state';
export * from './components/category-filter/logical-filter-operator';
export { FlipTileState } from './components/flip-tile/flip-tile-state';
export * from './components/modal/modal-utils';
export * from './components/toast/toast-utils';
export * from './components/tree-item/default-tree-item';
export * from './components/tree/tree-model';
export * from './components/upload/upload-file-state';
export { NotificationColor } from './components/utils/notification-color';
export {
  convertToAbbreviationString,
  convertToRemString,
} from './components/utils/rwd.util';
export * from './components/utils/theme-switcher';

declare module './components' {
  export namespace Components {
    export type IxIcon = IxIconsComponents.IxIcon;
  }
}

declare module './components' {
  export namespace JSX {
    export type IxIcon = IxIconsJSX.IxIcon;
  }
}
