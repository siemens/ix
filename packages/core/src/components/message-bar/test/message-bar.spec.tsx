/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { newSpecPage } from '@stencil/core/testing';
import { fireEvent } from '@testing-library/dom';
import { MessageBar } from '../message-bar';

describe('ix-message-bar', () => {
  let page: any;
  let messageBar: HTMLIxMessageBarElement;
  let closeButton: HTMLIxIconButtonElement;

  beforeEach(async () => {
    page = await newSpecPage({
      components: [MessageBar],
      html: '<ix-message-bar type="danger"></ix-message-bar>',
    });

    messageBar = document.querySelector('ix-message-bar');
    closeButton = messageBar.shadowRoot.querySelector(
      '[data-testid="close-btn"]'
    );
  });

  it('renders', async () => {
    expect(page.root).toMatchSnapshot();
  });

  it('closes the alert message bar', async (done) => {
    messageBar.addEventListener('closedChange', () => {
      done();
    });
    fireEvent.click(closeButton);
  });

  it('emits an event when the message is dismissed', async (done) => {
    const mockCallback = jest.fn(() => {
      done();
    });
    window.addEventListener('closedChange', mockCallback);

    fireEvent.click(closeButton);
    await page.waitForChanges();

    window.removeEventListener('closedChange', mockCallback);
  });
});
