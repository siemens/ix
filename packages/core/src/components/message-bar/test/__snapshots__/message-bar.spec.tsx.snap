// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ix-message-bar renders 1`] = `
<ix-message-bar type="danger">
  <mock:shadow-root>
    <div class="danger message-container" role="alert">
      <ix-icon color="color-alarm" name="error"></ix-icon>
      <div class="message-content">
        <slot></slot>
      </div>
      <ix-icon-button data-testid="close-btn" ghost="" icon="close" size="24"></ix-icon-button>
    </div>
  </mock:shadow-root>
</ix-message-bar>
`;
